### ISSUE TITLE
Compiling given CGAL example on polyfit

### ISSUE THREAD

**Issue by Mzzzzzzzzz** [CGAL/cgal#5221]
## Issue Details

I use visual studio2019 running the example from https://doc.cgal.org/latest/Polygonal_surface_reconstruction/Polygonal_surface_reconstruction_2polyfit_example_without_input_planes_8cpp-example.html

I have linked with libgmp-10.lib,libmpfr-4.lib,libscip.lib and I got the error
Severity    Code    Description Project File    Line    Suppression State
Error   LNK2019 unresolved external symbol __gmpn_neg_n referenced in function "private: static struct CGAL::Mpzf __cdecl CGAL::Mpzf::aors(struct CGAL::Mpzf const &,struct CGAL::Mpzf const &,int)" (?aors@Mpzf@CGAL@@CA?AU12@AEBU12@0H@Z) cgal-polyfit

the libgmp-10.lib,libmpfr-4.lib are located in the C:\dev\CGAL-5.1.1\auxiliary\gmp

## Source Code
```c++
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/IO/read_xyz_points.h>
#include <CGAL/IO/Writer_OFF.h>
#include <CGAL/property_map.h>
#include <CGAL/Surface_mesh.h>
#include <CGAL/Shape_detection/Efficient_RANSAC.h>
#include <CGAL/Polygonal_surface_reconstruction.h>


#define CGAL_USE_SCIP  // defined (or not) by CMake scripts, do not define by hand

#include <CGAL/SCIP_mixed_integer_program_traits.h>
typedef CGAL::SCIP_mixed_integer_program_traits<double>                        MIP_Solver;

//#elif defined(CGAL_USE_GLPK)  // defined (or not) by CMake scripts, do not define by hand
//
//#include <CGAL/GLPK_mixed_integer_program_traits.h>
//typedef CGAL::GLPK_mixed_integer_program_traits<double>                        MIP_Solver;
//
//#endif


#if defined(CGAL_USE_GLPK) || defined(CGAL_USE_SCIP)

#include <CGAL/Timer.h>

#include <fstream>


typedef CGAL::Exact_predicates_inexact_constructions_kernel                        Kernel;

typedef Kernel::Point_3                                                                                                Point;
typedef Kernel::Vector_3                                                                                        Vector;

// Point with normal, and plane index
typedef boost::tuple<Point, Vector, int>                                                        PNI;
typedef std::vector<PNI>                                                                                        Point_vector;
typedef CGAL::Nth_of_tuple_property_map<0, PNI>                                                Point_map;
typedef CGAL::Nth_of_tuple_property_map<1, PNI>                                                Normal_map;
typedef CGAL::Nth_of_tuple_property_map<2, PNI>                                                Plane_index_map;

typedef CGAL::Shape_detection::Efficient_RANSAC_traits<Kernel, Point_vector, Point_map, Normal_map>     Traits;

typedef CGAL::Shape_detection::Efficient_RANSAC<Traits>             Efficient_ransac;
typedef CGAL::Shape_detection::Plane<Traits>                                                Plane;
typedef CGAL::Shape_detection::Point_to_shape_index_map<Traits>     Point_to_shape_index_map;

typedef        CGAL::Polygonal_surface_reconstruction<Kernel>                                Polygonal_surface_reconstruction;
typedef CGAL::Surface_mesh<Point>                                                                        Surface_mesh;

/*
* This example first extracts planes from the input point cloud
* (using RANSAC with default parameters) and then reconstructs
* the surface model from the planes.
*/

int main()
{
    Point_vector points;

    // Loads point set from a file.
    const std::string input_file("data/cube.pwn");
    std::ifstream input_stream(input_file.c_str());
    if (input_stream.fail()) {
        std::cerr << "failed open file \'" << input_file << "\'" << std::endl;
        return EXIT_FAILURE;
    }
    std::cout << "Loading point cloud: " << input_file << "...";

    CGAL::Timer t;
    t.start();
    if (!input_stream ||
        !CGAL::read_xyz_points(input_stream,
            std::back_inserter(points),
            CGAL::parameters::point_map(Point_map()).normal_map(Normal_map())))
    {
        std::cerr << "Error: cannot read file " << input_file << std::endl;
        return EXIT_FAILURE;
    }
    else
        std::cout << " Done. " << points.size() << " points. Time: " << t.time() << " sec." << std::endl;

    // Shape detection
    Efficient_ransac ransac;
    ransac.set_input(points);
    ransac.add_shape_factory<Plane>();

    std::cout << "Extracting planes...";
    t.reset();
    ransac.detect();

    Efficient_ransac::Plane_range planes = ransac.planes();
    std::size_t num_planes = planes.size();

    std::cout << " Done. " << num_planes << " planes extracted. Time: " << t.time() << " sec." << std::endl;

    // Stores the plane index of each point as the third element of the tuple.
    Point_to_shape_index_map shape_index_map(points, planes);
    for (std::size_t i = 0; i < points.size(); ++i) {
        // Uses the get function from the property map that accesses the 3rd element of the tuple.
        int plane_index = get(shape_index_map, i);
        points[i].get<2>() = plane_index;
    }


    std::cout << "Generating candidate faces...";
    t.reset();

    Polygonal_surface_reconstruction algo(
        points,
        Point_map(),
        Normal_map(),
        Plane_index_map()
    );

    std::cout << " Done. Time: " << t.time() << " sec." << std::endl;


    Surface_mesh model;

    std::cout << "Reconstructing...";
    t.reset();

    if (!algo.reconstruct<MIP_Solver>(model)) {
        std::cerr << " Failed: " << algo.error_message() << std::endl;
        return EXIT_FAILURE;
    }

    const std::string& output_file("data/cube_result.off");
    std::ofstream output_stream(output_file.c_str());
    if (output_stream && CGAL::write_off(output_stream, model)) {
        // flush the buffer
        output_stream << std::flush;
        std::cout << " Done. Saved to " << output_file << ". Time: " << t.time() << " sec." << std::endl;
    }
    else {
        std::cerr << " Failed saving file." << std::endl;
        return EXIT_FAILURE;
    }


    // Also stores the candidate faces as a surface mesh to a file
    Surface_mesh candidate_faces;
    algo.output_candidate_faces(candidate_faces);
    const std::string& candidate_faces_file("data/cube_candidate_faces.off");
    std::ofstream candidate_stream(candidate_faces_file.c_str());
    if (candidate_stream && CGAL::write_off(candidate_stream, candidate_faces)) {
        // flush the buffer
        output_stream << std::flush;
        std::cout << "Candidate faces saved to " << candidate_faces_file << "." << std::endl;
    }

    return EXIT_SUCCESS;
}


#else

int main(int, char**)
{
    std::cerr << "This test requires either GLPK or SCIP.\n";
    return EXIT_SUCCESS;
}

#endif  // defined(CGAL_USE_GLPK) || defined(CGAL_USE_SCIP)
```
## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits):Windows 64bits
* Compiler:visual studio2019
* Release or debug mode:debug
* Specific flags used (if any):
* CGAL version:5.1.1
* Boost version:1.7.2
* Other libraries versions if used (Eigen, TBB, etc.):SCIPOptSuite 7.0.1

**Comment by danston** [comment]
Hi, @Mzzzzzzzzz Did you change the cmake for any reason? As far as I understand this error, it is related to how you are linking with the external libraries. Do you use vcpkg as explained [here](https://doc.cgal.org/latest/Manual/windows.html)?

**Comment by Mzzzzzzzzz** [comment]
i have installed cgal by exe and vcpkg, but i can not find the libgmp-10.lib,libmpfr-4.lib in the directory  D:\vcpkg-master\vcpkg-master\installed\x86-windows\lib ,so i add the directory of cgal  installed by exe and the path is  C:\dev\CGAL-5.1.1\auxiliary\gmp

**Comment by janetournois** [comment]
Hi @Mzzzzzzzzz ,

I have just installed CGAL using vcpkg on my machine, and I have in the folder `vcpkg\installed\x64-windows\lib` the libraries `mpfr.lib` and `mpir.lib` available (everything was just installed).
Note that with vcpkg on windows, you should provide MPIR as an implementation of GMP.

So : 
`GMP_INCLUDE_DIR=vcpkg/installed/x64-windows/include`,
`GMP_LIBRARIES=vcpkg/installed/x64-windows/lib/mpir.lib`

Would this work for you?

**Comment by Mzzzzzzzzz** [comment]
I added the mpfr.lib and mpir.lib to the linker,but I still get the same error,so I wonder maybe the _gmpn_neg_n is not included in the GMP?

**Comment by janetournois** [comment]
Have you also updated both include paths, and library dirs, in CMake? (with everything from vcpkg and nothing from the "other" installation of CGAL)

**Comment by Mzzzzzzzzz** [comment]
if i remove the libgmp-10.lib from the linker ,i will get such a lot link error about gmp

**Comment by brandonvno** [comment]
Hi @Mzzzzzzzzz , I have the same problem just as you and I solved it by adding `D:\dev\CGAL-5.1.1\auxiliary\gmp\lib\libgmp-10.lib` and `E:\SourceCode\vcpkg\installed\x64-windows\lib\mpir.lib` to my project in vs2019.  
And thanks for your thought! @janetournois 


**Environment:**
- Operating system (Windows/Mac/Linux, 32/64 bits):Windows 64bits
- Compiler:visual studio2019
- Release or debug mode: release
- Specific flags used (if any):
- CGAL version:5.1.1(installer and vcpkg)
- Boost version:1.7.4

**Comment by Mzzzzzzzzz** [comment]
thanks a lot！ i rerun the program and success!

**Comment by FuzhangHe** [comment]
> Hi @Mzzzzzzzzz ,
> 
> I have just installed CGAL using vcpkg on my machine, and I have in the folder `vcpkg\installed\x64-windows\lib` the libraries `mpfr.lib` and `mpir.lib` available (everything was just installed). Note that with vcpkg on windows, you should provide MPIR as an implementation of GMP.
> 
> So : `GMP_INCLUDE_DIR=vcpkg/installed/x64-windows/include`, `GMP_LIBRARIES=vcpkg/installed/x64-windows/lib/mpir.lib`
> 
> Would this work for you?

Hallo, @janetournois ,

I use Vcpkg to install cgal, and have the same problem. i have automaticly installed GMP.
What do you mean by "you should provide MPIR as an implementation of GMP" ? how can i do it ?
if i try to install MPIR,  i get the error:
     " Can't build mpir if gmp is installed.  Please remove gmp, and try to  install mpir again if you need it."

best regards

**Comment by janetournois** [comment]
Hello @FuzhangHe,
if you have GMP, why not providing it directly through `GMP_INCLUDE_DIR` and `GMP_LIBRARIES`, in CMake?

**Comment by FuzhangHe** [comment]
> Hello @FuzhangHe, if you have GMP, why not providing it directly through `GMP_INCLUDE_DIR` and `GMP_LIBRARIES`, in CMake?

Hi, Jane,
Thank you for you replay, I am new in C++ and CGAL.
To make it clear, i write here what i have done und what is my problem in detail:

- I have installed CGAL with VCPKG, just as the manual. Then I compiled the example "Triangulation_2" with cmake-gui, it all works without error.

- Also the Tutorial Hello World also works for me in Visual Studio 2019. [https://doc.cgal.org/latest/Manual/tutorial_hello_world.html](url)

- Then I try the nef_3_simple.cpp in Visual Studio 2019. I create a new project and copy this file in the "Source Files" Folder. In the project properties, i add ...\dev\vcpkg\installed\x64-windows\lib and ...\dev\vcpkg\installed\x64-windows\lib to the Linker--> General; add gmp.lib and mpfr.lib to Linker--> Input. Then I got the error: gmp.dll not found.

- I just added libgmp-10.lib, libmpfr-4.lib, libgmp-10.dll, libmpfr-4.dll to my project lib folder "...\CGAL_HelloWorld\lib", it seems to work.
